[WGSL] Start implementing the codegen

tadeuzagallo opened the pull request:
#### 06f047e2888cd627f37a1f74ade3447292e54021
<pre>
[WGSL] Start implementing the codegen
<a href="https://bugs.webkit.org/show_bug.cgi?id=250555">https://bugs.webkit.org/show_bug.cgi?id=250555</a>
&lt;rdar://problem/104217343&gt;

Reviewed by Myles C. Maxfield.

A initial naive codegen implementation, just enough to show the red triangle on the screen.

* Source/WebGPU/WGSL/Metal/MetalCodeGenerator.cpp: Added.
(WGSL::Metal::metalCodePrologue):
(WGSL::Metal::dumpMetalCodeIfNeeded):
(WGSL::Metal::generateMetalCode):
* Source/WebGPU/WGSL/Metal/MetalCodeGenerator.h: Added.
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp: Added.
(WGSL::Metal::FunctionDefinitionWriter::FunctionDefinitionWriter):
(WGSL::Metal::FunctionDefinitionWriter::visit):
(WGSL::Metal::emitMetalFunctions):
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.h: Added.
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser&lt;Lexer&gt;::parseStructMember):
(WGSL::Parser&lt;Lexer&gt;::parseFunctionDecl):
* Source/WebGPU/WGSL/WGSL.cpp:
(WGSL::prepare):
* Source/WebGPU/WGSL/WGSL.h:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/ComputePipeline.mm:
(WebGPU::createLibrary):
(WebGPU::Device::createComputePipeline):
* Source/WebGPU/WebGPU/ShaderModule.h:
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::ShaderModule::ast const):
* Websites/webkit.org/demos/webgpu/scripts/hello-triangle.js:
(async helloTriangle):

Canonical link: <a href="https://commits.webkit.org/259070@main">https://commits.webkit.org/259070@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/33bfadb53c7083a116a16fbc4557f6cdc6d31465

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103725 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12841 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36670 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112952 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13863 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3735 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95979 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112085 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109496 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10688 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93750 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38411 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92512 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25356 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80061 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6202 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26745 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6376 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3275 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12359 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46271 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6239 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8136 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Websites/webkit.org/demos/webgpu/scripts/hello-triangle.js:
```diff
@@ -30,17 +30,17 @@ async function helloTriangle() {
 /*
```
did you want to uncomment the wgsl source and remove (or comment out) the metal source from this file?

tadeuzagallo replied:
I did that locally, but I was waiting until we could display the colors to replace the metal source

litherum commented on Source/WebGPU/WGSL/Metal/MetalCodeGenerator.cpp:
```diff
@@ -0,0 +1,71 @@
+static void dumpMetalCodeIfNeeded(StringBuilder& stringBuilder)
+{
+    if (dumpMetalCode) {
+        dataLogLn("Generated Metal code: ");
```
Can we use WTFLogAlways?

tadeuzagallo replied:
Sure, I just ported this from an old implementation. Any reason to use one over the other? I don't think I've used `WTFLogAlways` before, and it wasn't obvious from the implementation.

djg replied:
I heard that `dataLogLn` was a favourite of the JSC team.

litherum replied:
WTFLog is more fully featured; there are channels that can be toggled individually... that's what we use in WebCore.

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -0,0 +1,419 @@
+    void visit(AST::AbstractFloatLiteral&) override;
+    void visit(AST::AbstractIntLiteral&) override;
+    void visit(AST::ArrayAccess&) override;
+    // virtual void visit(AST::BoolLiteral&);
```
We usually don't leave code commented out; we either uncomment it so it's used, or delete the line altogether. Is there a reason we're leaving it here?

tadeuzagallo replied:
Fair enough, I just copied all the methods from the visitor, since we'll need to implement them later, but I'll just remove for now.

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -0,0 +1,419 @@
+
+void FunctionDefinitionWriter::visit(AST::BuiltinAttribute& builtin)
+{
+    if (builtin.name() == "vertex_index"_s) {
+        m_stringBuilder.append("[[vertex_id]]");
+        return;
+    }
+
+    if (builtin.name() == "position"_s) {
+        m_stringBuilder.append("[[position]]");
+        return;
+    }
```
Eventually, we'll want some kind of trie, or a perfect hash, or something, right? Maybe add a FIXME?

tadeuzagallo replied:
Will add a FIXME.

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -0,0 +1,419 @@
+{
+    switch (stage.stage()) {
+    case AST::StageAttribute::Stage::Vertex:
+        m_stringBuilder.append("vertex");
```
The modern MSL syntax for this is `[[vertex]]`, `[[fragment]]` and `[[compute]]`.

djg replied:
Which didn't actually work with something I tried.

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -0,0 +1,419 @@
+{
+    ASSERT(type.maybeElementType());
+    ASSERT(type.maybeElementCount());
+    m_stringBuilder.append("array<");
```
Don't we want to be explicit and say `metal::array`?

tadeuzagallo replied:
We declare `using namespace metal` in the codegen, and I believe array is a reserved keyword in WGSL, so I don't think it should be necessary?

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -0,0 +1,419 @@
+
+void FunctionDefinitionWriter::visit(AST::NamedType& type)
+{
+    if (type.name() == "i32"_s)
+        m_stringBuilder.append("int");
+    else if (type.name() == "f32"_s)
+        m_stringBuilder.append("float");
+    else if (type.name() == "u32"_s)
+        m_stringBuilder.append("unsigned");
+    else
+        m_stringBuilder.append(type.name());
```
Ditto about using a better data structure, rather than cascaded `if`s

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -0,0 +1,419 @@
+        m_stringBuilder.append(", 4>");
+        break;
+    case AST::ParameterizedType::Base::Mat2x2:
+        // TODO
```
I think usually we say FIXME and link to a bug

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -0,0 +1,419 @@
+void FunctionDefinitionWriter::visit(AST::ParameterizedType& type)
+{
+    switch (type.base()) {
+    case AST::ParameterizedType::Base::Vec2:
```
Very curious about your opinions on having `Vec2` be explicit AST nodes, rather than just regular types.

On one hand, the WGSL grammar explicitly mentions them (I think because they take `<>` parameters, and the editors want special handling for that), but on the other hand, if we handle all types systemically, it removes the need for special casing them throughout our code.

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -0,0 +1,419 @@
+void FunctionDefinitionWriter::visit(AST::CallableExpression& call)
+{
+    bool first = true;
+    if (call.target().kind() == AST::Node::Kind::ArrayType) {
```
There's not a more elegant way of handling this? it seems weird for code to indicate we're trying to call an array.

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -0,0 +1,419 @@
+        for (auto& argument : call.arguments()) {
+            if (!first)
+                m_stringBuilder.append(", ");
+            visit(argument);
```
I think we need to hoist these, to make sure order of execution is well-defined. Consider `f(g(), h())`

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -0,0 +1,419 @@
+    case AST::UnaryOperation::Negate:
+        m_stringBuilder.append("-");
+    }
+    visit(unary.expression());
```
I think we should spray parentheses everywhere, to make sure order of operations is preserved.

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -0,0 +1,419 @@
+
+void FunctionDefinitionWriter::visit(AST::IdentifierExpression& identifier)
+{
+    m_stringBuilder.append(identifier.identifier());
```
We need to mangle identifiers for security. No byte from the source program should end up in the destination program. Maybe add a FIXME?

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -0,0 +1,419 @@
+void FunctionDefinitionWriter::visit(AST::StructureAccess& access)
+{
+    visit(access.base());
+    m_stringBuilder.append(".", access.fieldName());
```
Ditto about mangling, this time for field names.

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -0,0 +1,419 @@
+
+void FunctionDefinitionWriter::visit(AST::AbstractIntLiteral& literal)
+{
+    m_stringBuilder.append(literal.value());
```
Are we sure this will serialize correctly, in every case?

tadeuzagallo replied:
No, I don't think it will. I'll add a FIXME for now.

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -0,0 +1,419 @@
+
+    AST::Visitor::visit(statement);
+
+    if (statement.kind() != AST::Node::Kind::CompoundStatement)
```
What is an example of a compound statement?

tadeuzagallo replied:
A compound statement is a list of statements, e.g. the body of a function

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.h:
```diff
@@ -0,0 +1,44 @@
+};
+RenderMetalFunctionEntryPoints emitMetalFunctions(StringBuilder&, AST::ShaderModule&);
+
+} } // namespace WGSL::Metal
```
Nit: can we put the `}` on different lines?

tadeuzagallo replied:
I just realized this might be a JSC-ism, doesn't seem to be used in other parts of WK

litherum commented on Source/WebGPU/WGSL/WGSL.cpp:
```diff
@@ -54,19 +55,19 @@ SuccessfulCheck::SuccessfulCheck(Vector<Warning>&& messages, UniqueRef<AST::Shad
 
 SuccessfulCheck::~SuccessfulCheck() = default;
 
-PrepareResult prepare(const AST::ShaderModule& ast, const HashMap<String, PipelineLayout>& pipelineLayouts)
```
I think we should try to keep these const. One of the lessons I've learned from the LFC work we've been doing in WebKit is that it's really important to be able to have strict contracts about what is populated when and by which thing; and the C++ type system is a great way of enforcing it. Conceptually, the AST is the input to the preparation function, and conceptually, preparation shouldn't modify the AST; if we want to fill in some previously-unpopulated fields inside the AST, I think we should use other side data structures for those. Having strong contracts around this stuff makes the code way more easy to hack on; you don't have to manually figure out which field was written by which stage.

litherum commented on Websites/webkit.org/demos/webgpu/scripts/hello-triangle.js:
```diff
@@ -30,17 +30,17 @@ async function helloTriangle() {
-                     @vertex fn main(@builtin(vertex_index) VertexIndex: u32) -> @builtin(position) vec4<f32>
+                     @vertex fn vsmain(@builtin(vertex_index) VertexIndex: u32) -> @builtin(position) vec4<f32>
                      {
-                         var pos = array<vec2<f32>, 3>(
+                         var pos: array<vec2<f32>, 3> = array<vec2<f32>, 3>(
                              vec2<f32>( 0.0,  0.5),
                              vec2<f32>(-0.5, -0.5),
                              vec2<f32>( 0.5, -0.5)
                          );
                          return vec4<f32>(pos[VertexIndex], 0.0, 1.0);
                      }
 
-                     @fragment fn main() -> @location(0) vec4<f32>
+                     @fragment fn fsmain() -> @location(0) vec4<f32>
                      {
```
What would it take to make real tests, rather than just these demo files? I think we should either A) use existing tests from the CTS from upstream, or B) write our own real tests and contribute them to the upstream CTS.

tadeuzagallo replied:
I'll investigate that, I honestly haven't started running the real tests yet, I was just playing around with the demo for now, but I agree we should start testing it properly.

mwyrzykowski replied:
I agree we should do this too but it is probably outside the scope of this PR.

@litherum do you know who is responsible for maintaining [the CTS](https://github.com/gpuweb/cts)? Is that any of the gpuweb contributors? I think its non-functional at the moment

litherum replied:
We have a snapshot of the CTS already in WebKit that runs on every commit.

Yes, the maintainers are the WebGPU contributors.

tadeuzagallo committed:
[WGSL] Start implementing the codegen
https://bugs.webkit.org/show_bug.cgi?id=250555
<rdar://problem/104217343>

Reviewed by Myles C. Maxfield.

A initial naive codegen implementation, just enough to show the red triangle on the screen.

* Source/WebGPU/WGSL/Metal/MetalCodeGenerator.cpp: Added.
(WGSL::Metal::metalCodePrologue):
(WGSL::Metal::dumpMetalCodeIfNeeded):
(WGSL::Metal::generateMetalCode):
* Source/WebGPU/WGSL/Metal/MetalCodeGenerator.h: Added.
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp: Added.
(WGSL::Metal::FunctionDefinitionWriter::FunctionDefinitionWriter):
(WGSL::Metal::FunctionDefinitionWriter::visit):
(WGSL::Metal::emitMetalFunctions):
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.h: Added.
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser<Lexer>::parseStructMember):
(WGSL::Parser<Lexer>::parseFunctionDecl):
* Source/WebGPU/WGSL/WGSL.cpp:
(WGSL::prepare):
* Source/WebGPU/WGSL/WGSL.h:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/ComputePipeline.mm:
(WebGPU::createLibrary):
(WebGPU::Device::createComputePipeline):
* Source/WebGPU/WebGPU/ShaderModule.h:
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::ShaderModule::ast const):
* Websites/webkit.org/demos/webgpu/scripts/hello-triangle.js:
(async helloTriangle):

Canonical link: https://commits.webkit.org/259070@main
